pkg/pgmodel/migrations/sql/1_base_schema.up.sql

```diff
@@ -170,6 +186,55 @@ END
 $func$
 LANGUAGE PLPGSQL VOLATILE PARALLEL SAFE;
 
+--Creates a new label_key row for a given key.
+--This uses up some sequences so should only be called
+--If the table does not yet exist.
+CREATE OR REPLACE FUNCTION SCHEMA_CATALOG.create_label_key(
+        new_key TEXT, OUT id INT, OUT value_column_name NAME, OUT id_column_name NAME
+)
+AS $func$
+DECLARE
+  new_id int;
+BEGIN
+new_id = nextval(pg_get_serial_sequence('SCHEMA_CATALOG.label_key','id'))::int;
+LOOP
+    INSERT INTO SCHEMA_CATALOG.label_key (id, key, value_column_name, id_column_name)
+        SELECT  new_id,
+                new_key,
+                SCHEMA_CATALOG.pg_name_unique(new_key, new_id::text),
+                SCHEMA_CATALOG.pg_name_unique(new_key || '_id', format('%s_id', new_id))
+    ON CONFLICT DO NOTHING
+    RETURNING SCHEMA_CATALOG.label_key.id, SCHEMA_CATALOG.label_key.value_column_name, SCHEMA_CATALOG.label_key.id_column_name
+    INTO id, value_column_name, id_column_name;
+    -- under high concurrency the insert may not return anything, so try a select and loop
+    -- https://stackoverflow.com/a/15950324
+    EXIT WHEN FOUND;
+
+    SELECT lk.id, lk.value_column_name, lk.id_column_name
+    INTO id, value_column_name, id_column_name
+    FROM SCHEMA_CATALOG.label_key lk
+    WHERE key = new_key;
+
+    EXIT WHEN FOUND;
+END LOOP;
+END
+$func$
+LANGUAGE PLPGSQL VOLATILE PARALLEL SAFE;
+
+--Get a label key row if one doesn't yet exist.
+CREATE OR REPLACE FUNCTION SCHEMA_CATALOG.get_or_create_label_key(
+        key TEXT, OUT id INT, OUT value_column_name NAME, OUT id_column_name NAME)
+AS $func$
+   SELECT id, value_column_name, id_column_name
+   FROM SCHEMA_CATALOG.label_key lk
+   WHERE lk.key = get_or_create_label_key.key
+   UNION ALL
+   SELECT *
+   FROM SCHEMA_CATALOG.create_label_key(get_or_create_label_key.key)
```

Comment:
I'm not sure how limit works with this, it might call the create_label_key function because it's marked volatile and therefore needs to be called even if the other is found have you confirmed that this works right?

Reply:
Yes I have confirmed